torch/utils/data/sharding.py

```diff
@@ -0,0 +1,22 @@
+import torch.utils.data.graph
+
+
+def apply_sharding(datapipe, num_of_instances, instance_id):
+    graph = torch.utils.data.graph.traverse(datapipe)
+
+    def traverse_graph(graph):
+        results = set()
+        for datapipe, sub_graph in graph.items():
+            results.add(datapipe)
+            sub_items = traverse_graph(sub_graph)
+            for item in sub_items:
+                results.add(item)
+        return results
+
+    all_pipes = traverse_graph(graph)
+
+    for pipe in all_pipes:
+        if hasattr(pipe, 'is_shardable'):
+            if pipe.is_shardable():
+                if hasattr(pipe, 'apply_sharding'):
+                    pipe.apply_sharding(num_of_instances, instance_id)
```

Comment:
I have a question here. Current test only takes one `shardable` datapipe. What if we have multiple DataPipe within pipeline, we should only make sharding happens once, right? If that is the case, considering the performance, sharding should happen as early as possible.

Reply:
Good question, potentially we might have situation when A->B->C->D-> becomes 2 shards of B and 4 shards of C. But for sake of simple functionality at this particular moment, I will add exception if we shard more than once.